### PR TITLE
WIP feature(events): Allow class name as event/hook handler

### LINKED ISSUE #8077: Allow class name as event/hook handler
Idea by @ewinslow:

``` php
namespace MyPlugin\Events;
class UserLogin {
  const KLASS = __CLASS__;
  public function __invoke($e, $t, \ElggUser $user) {
    // logic
  }
}

elgg_register_event_handler('login', 'user', UserLogin::KLASS, 100);

elgg_unregister_event_handler('login', 'user', UserLogin::KLASS);

// PHP 5.5+
elgg_register_event_handler('login', 'user', UserLogin::class, 100);
```

Internally the handler would just be the string `MyPlugin\Events\UserLogin`, which could be a function name, so we would need to first check function_exists(). If that fails we ask the DI system for an object of that type.

### PR DESCRIPTION
**Pull request by mrclay**
Blocked on:
- [ ] Wait until we can bump to PHP 5.5. `::KLASS` is too hacky
- [ ] Wait until we can have a proper DI system, then this actually becomes more useful...

Fixes #8077

### CODE REVIEW

File: engine/classes/Elgg/PluginHooksService.php [hunk 1]
```diff
@@ -25,11 +25,35 @@ public function trigger($hook, $type, $params = null, $returnvalue = null) {
...
+
+				$class = ltrim($callback, '\\');
+
+				$cached = _elgg_services()->config->get("object:$class");
+				if (!$cached) {
+					$factory = _elgg_services()->config->get("class_factory:$class");
+					if ($factory) {
+						$cached = call_user_func($factory, $class);
+					} else {
+						$cached = new $class();
```
Comment (ewinslow) [inline engine/classes/Elgg/PluginHooksService.php]: I think this should be the only thing we support right now. Then when we offer DI we can use `$injector->instantiate($class);` or whatever the equivalent is.

File: mod/htmlawed/classes/HtmlawedPlugin/ValidateInputHook.php [hunk 2]
```diff
@@ -0,0 +1,141 @@
+<?php
+namespace HtmlawedPlugin;
```
Comment (ewinslow) [inline mod/htmlawed/classes/HtmlawedPlugin/ValidateInputHook.php]: We will need to decide on namespacing standards for plugins.

My thought is `Vendor\Plugin\Class`, so in this case `Elgg\Htmlawed\ValidateInputHook`

File: mod/htmlawed/start.php [hunk 3]
```diff
@@ -23,136 +29,6 @@ function htmlawed_init() {
 }
 
 /**
- * htmLawed filtering of data
```
Comment (ewinslow) [inline mod/htmlawed/start.php]: I like this a lot! Not so much bloat in start.php anymore...

Comment (ewinslow) [inline mod/htmlawed/start.php]: Automatic lazy-loading too, sweet...

File: mod/htmlawed/start.php [hunk 4]
```diff
@@ -14,7 +14,13 @@
  * Initialize the htmlawed plugin
  */
 function htmlawed_init() {
-	elgg_register_plugin_hook_handler('validate', 'input', 'htmlawed_filter_tags', 1);
+
+	// just to show off the factory feature
+	elgg_register_class_factory(\HtmlawedPlugin\ValidateInputHook::KLASS, function ($class) {
```
Comment (ewinslow) [inline mod/htmlawed/start.php]: This PR seems to mix DI with the class-as-handler feature. Can we just limit it to class-as-handler and do a "proper" DI implementation later?

**Comment by ewinslow** [comment]
So great! Thank you for starting on this!!

File: mod/htmlawed/start.php (hunk 3 above)
Comment (mrclay) [inline mod/htmlawed/start.php]: Well, no lazy loading if we use `::KLASS`

**Comment by mrclay** [comment]
@ewinslow updated to remove factory stuff

File: mod/htmlawed/start.php (hunk 3 above)
Comment (ewinslow) [inline mod/htmlawed/start.php]: Yes, not as lazy as we'd like anyhow

File: engine/classes/Elgg/EventsService.php [hunk 5]
```diff
@@ -44,11 +44,29 @@ public function trigger($event, $type, $object = null, array $options = array())
 		$args = array($event, $type, $object);
 
 		foreach ($events as $callback) {
+
+			$callback_copy = $callback;
+
+			if (is_string($callback)
+					&& false !== strpos($callback, '::')
+					&& !function_exists($callback)
+					&& class_exists($callback)) {
```
Comment (juho-jaakkola) [inline engine/classes/Elgg/EventsService.php]: This isn't very readable. How about moving these checks into a separate method? (```isClassCallback()``` or something)

Comment (mrclay) [inline engine/classes/Elgg/EventsService.php]: I'm just getting buy in for the feature before putting much more effort in.

File: mod/htmlawed/classes/HtmlawedPlugin/ValidateInputHook.php (hunk 2 above)
Comment (mrclay) [inline mod/htmlawed/classes/HtmlawedPlugin/ValidateInputHook.php]: `Elgg\Mod\` for plugins would eliminate chance of conflict with system classes. Creating these directory trees is annoying. We need an elgg-cli.

**Comment by mrclay** [comment]
I think this feature isn't going to solve any immediate problems and maybe should wait for DI to improve and we can bump to PHP 5.5. It would also be nice to move the instantiation to getOrderedHandlers() maybe.

**Comment by ewinslow** [comment]
SGTM

**Comment by ewinslow** [comment]
Added as TODOs, will leave this alone until we have those.

**Comment by mrclay** [comment]
Replaced by https://github.com/Elgg/Elgg/pull/8399

**Pull request closed without merging**